### PR TITLE
Enable kubectl-get to strip managed fields

### LINKED ISSUE #90066: managedFields options are very verbose, making kubectl usage a bit
Recently (1.18?) a new field was added to configs, managedFields. This is incredibly verbose, which makes kubectl commands like `-oyaml` and `edit` painful.

For example, on a random deployment I have looked at, the entire `kubectl get deployment -oyaml` is 640 lines. Managed fields is over half of that, with 358 lines.

I am not sure if its feasible to somehow mitigate this without breaking server side apply or other features, but it would be useful if possible. One possible idea would be to just hide it from kubectl commands, but I am not sure the impact of that. The goal of this issue is mostly to bring awareness that this causes some user pain and see if we can improve that a bit.

Version:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T20:56:08Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by knight42**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Enable `kubectl get` to strip managed fields to make the output less verbose.

**Which issue(s) this PR fixes**:

xref #90066

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubectl: `kubectl get` will omit managed fields by default now. Users could set `--show-managed-fields` to true to show managedFields when the output format is either `json` or `yaml`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by luisdavim** [comment]
This helps but doesn't fix #90066

**Comment by knight42** [comment]
@luisdavim okay, I have updated this PR's description so that #90066 won't be closed.

**Comment by kwiesmueller** [comment]
I think I'd rather prefer an additional format for the -o flag than this flag.
`edit` is already fixed and only `get` should be remaining, this flag here is very generic.

There were a few suggestions already, something like `-o simpleyaml`.

Providing a format that does not show managedFields is keeping the current formats like they are.

We should not drop any fields on get without it being an active decision by the user, as currently people are used to get giving them a full dump of the object.

Also I think we should try to solve this on the Server-Side and then use the flag in the client.

**Comment by knight42** [comment]
> currently people are used to get giving them a full dump of the object.

@kwiesmueller I think this is a good point, thanks for pointing it out.

And if there is anything I could do now please let me know, I am willing to help! e.g. adding a new format `simpleyaml`

/hold

**Comment by kwiesmueller** [comment]
Right now, the task is unassigned. If you'd like to help, the best thing would be to stop by in the wg-api-expression Slack channel.

I think we still need to discuss what the final solution will be, but help is always appreciated.

As far as I remember there still is the question if we're able to filter the field at the server (some possible limitations with watch etc.).

**Comment by soltysh** [comment]
SSA will be brought into discussion during next-week's SIG-CLI, we'll go over this PR as well and we'll try to make the final call.
Thanks @knight42 for holding with it for now :)

**Comment by julianvmodesto** [comment]
Given that we're planning for a server-side header to hide managed fields, it does seem like a dedicated flag would make sense and it's smaller than defining behavior for a flag like -oshort.

But we should update the KEP with the server-side header and the kubectl flag plan since we want to get this in for this release.

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go:68]: I'd possibly name this variable `show-managed-fields`, and default it to false. That seems to be the general consensus so far.

File: staging/src/k8s.io/cli-runtime/pkg/printers/json_test.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/json_test.go:144]: I'm curious if "diff" will work the same way. i.e. will it diff the managed fields or will they be dropped anyway? Would you mind double checking how we print objects for diff?

File: staging/src/k8s.io/cli-runtime/pkg/printers/json.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/json.go:155]: is it possible to build a "ManagedFieldsStripper" Printer, that will encapsulate another stripper but remove the ManagedFields from the object first? I think that'd be a slightly nicer design!

Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/json.go]: Yeah I think adding a new printer is better and I have added an `OmitManagedFields` printer similar to your proposal.

File: staging/src/k8s.io/cli-runtime/pkg/printers/json_test.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/json_test.go]: I have tried locally and it turns out that `diff` will diff the managed fields as well.

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go]: I have updated the flag as you said.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit.go:35]: it's unclear why this function ignores errors returned by `meta.Accessor`?

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:37]: I would drop this function altogether and compare the exact thing that you want as the output. In other words, I would make `testResourcePrinter` take the expected object that you think it will receive, and test that it does. Maybe something like this:

```golang
type testResourcePrinter struct {
    expected runtime.Object
}

func (p testResourcePrinter) PrintObj(o runtime.Object, w io.Writer) error {
    if !reflect.DeepEqual(o, p.expected) {
        return fmt.Errorf("%v is different from expected: %v", o, p.expected)   
    }
    return nil
}
```

Then you testcase only needs the input and expected output. Does that make sense?

File: staging/src/k8s.io/kubectl/pkg/cmd/get/get.go
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/get/get.go:272]: Why in `Get`? I think you can wired it up into https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go#L45-L59, and the flag can be added here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/json_yaml_flags.go#L63. I think it's better if the option is applied to all commands that print objects!

Unfortunately `kubectl diff` is a little different, we could take care of that right after.

**Review by apelisse — COMMENTED** [review]
That's awesome, a few minor comments but that looks really good!

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit.go:35]: `meta.Accessor` returning an error means the object does not implement `metav1.Object`, so I think it might be better to leave it untouched.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go]: Thanks, I think this is better.

File: staging/src/k8s.io/kubectl/pkg/cmd/get/get.go
Comment (knight42) [reply staging/src/k8s.io/kubectl/pkg/cmd/get/get.go]: Updated as you said.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit.go:41]: Do you need to return `o`?

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:32]: I generally avoid moving t around if I don't have too.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:37]: you could return the error here, and then you wouldn't need `t` in that object

**Review by apelisse — COMMENTED** [review]
That's very good and very close, minor nits left!

@soltysh, do you want to take a look?

Thanks!

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (apelisse) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:40]: These tests look much better now!

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit.go:41]: I return `o` here just for convenience, otherwise I have to change line 56 to
```go
obj = obj.DeepCopyObject()
omitManagedFields(obj)
```
we could save one line now.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:37]: `t` is required by `require.Equal`. `require.Equal` is useful when we need to compare objects, if the two objects are different, `require.Equal` would fail the test and print the diff which makes debugging easier.

Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go]: I have refactored the test a bit, could you take another look?

**Comment by apelisse** [comment]
Looks good to me. @soltysh PTAL, that goes with the KEP in https://github.com/kubernetes/enhancements/pull/2212

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (soltysh) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit.go:1]: I'd probably name this file `managedfields.go`.

**Review by soltysh — COMMENTED** [review]
One small nit about filename and this is good to go as is. Also please squash your changes.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit.go
Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit.go]: @soltysh I have updated the file name.

File: staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go
Comment (soltysh) [inline staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go:1]: @knight42 this one should match the rename, so `managedfields_test.go`

Comment (knight42) [reply staging/src/k8s.io/cli-runtime/pkg/printers/omit_test.go]: Updated

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Review by dougsland — COMMENTED** [review]
Little bit late in the game but LGTM
/lgtm

**Comment by alexec** [comment]
This is awesome. I hate managed fields.

**Comment by curusarn** [comment]
Is this going to be backported to older kubectl versions?

**Comment by thiagolsfortunato** [comment]
Until the Kubernetes upgrade to version 1.21+, I'm using `yq` to delete managed fields from the output.

```
$ kubectl get deploy api-foo -o yaml | yq e 'del(.metadata.managedFields)' -
```

**Comment by iam-veeramalla** [comment]
I am using kubectl -> 1.22.1

When I run the below commands, I still see the managed fields. 
`kubectl get service foo -o yaml`
`kubectl get --show-managed-feilds=false service foo  -o yaml `

**Comment by apelisse** [comment]
@iam-veeramalla If you have a bug please open a separate issue.

**Comment by navry** [comment]
@iam-veeramalla  you have typo there, `show-managed-fields` instead of `show-managed-feilds`

**Comment by iam-veeramalla** [comment]
> @iam-veeramalla  you have typo there, `show-managed-fields` instead of `show-managed-feilds`

My bad .. thanks

**Pull request merged**
